src/resolvers/Promotion.js

```diff
@@ -0,0 +1,267 @@
+const PromotionResolver = {
+  Query: {
+    promotions: async (root, { filter }, context, info) => {
+      /**
+       * - only allow logged in user
+       * - check filter validity
+       * - retrive based on filter
+       */
+      if (filter === null) {
+        throw new UserInputError("Invalid filter");
+      }
+
+      /**
+       * check filter validity
+       */
+      let parsedFilter;
+      try {
+        parsedFilter = JSON.parse(filter);
+      } catch (error) {
+        throw new ApolloError(error, 500);
+      }
+
+      /**
+       * retrieve based on filter
+       */
+      let promotions = await Promotion.find(parsedFilter);
+      promotions.map((promotion) => {
+        promotionConverter(promotion);
+      });
+
+      return promotions;
+    },
+    promotion: async (root, { id }, context, info) => {
+      /**
+       * - only allow logged in user
+       * - check id validity
+       * - retrive based on id
+       */
+
+      /**
+       * check id validity
+       */
+      if (id === null) {
+        throw new UserInputError("No ID provided");
+      }
+
+      /**
+       * validate id format
+       */
+      if (!mongoose.Types.ObjectId.isValid(id))
+        throw new UserInputError(`${id} is having an invalid format`);
+
+      /**
+       * retrive based on id
+       */
+      let promotion = await Promotion.findById(id);
+      promotion = promotionConverter(promotion);
+      promotion.id = promotion._id;
+
+      return promotion;
+    },
+  },
+  Mutation: {
+    createPromotion: async (root, { promotionInput }, context, info) => {
+      /**
+       * - only allow logged in Brand Owner to create promotions
+       * - throw if id provided
+       * - check if the promotion name is used
+       * - check if the date range is valid
+       * - insert promotion
+       */
+
+      /**
+       * check user type is Brand owner
+       */
+      if (context.user.type !== USER_TYPE.BRANDOWNER) {
+        throw AuthenticationError("Invalid Permission");
+      }
+
+      /**
+       * disallow id
+       */
+      if (promotionInput.id !== null) {
+        throw new UserInputError("ID not allowed while creating promotion");
+      }
+
+      /**
+       * - check validity of promotion name
+       * - disallow name if the promotion is still ongoing
+       */
+      const existingPromotion = await Promotion.findOne({
+        promotionNm: promotionInput.promotionNm,
+      });
+      if (
+        existingPromotion &&
+        new Date(existingPromotion.promotionEnd) > Date.now()
+      ) {
+        throw new ApolloError("Prmotion name alread taken");
+      }
+
+      /**
+       * date range validity check:
+       * - must start at least one week from now
+       * - end date > start date
+       */
+      const validStartDt =
+        new Date(promotionInput.promotionStart) >
+        new Date(Date.now() + 1000 * 60 * 60 * 24 * 7);
```

Comment:
I do not allow to create promotion if it is not one week from now. So that they have time to prepare for promotion. But this is unnecessary logic. If feels like wrong can remove.